pennylane/devices/default_clifford.py

```diff
@@ -531,46 +768,238 @@ def _measure_state(self, tableau_simulator, circuit, global_phase):
-    def _measure_density_matrix(tableau_simulator):
+    def _measure_density_matrix(tableau_simulator, wires):
         """Measure the density matrix from the state of simulator device."""
         state_vector = qml.math.array(tableau_simulator.state_vector(endian="big"))
-        return qml.math.einsum("i, j->ij", state_vector, state_vector)
+        return qml.math.reduce_dm(qml.math.einsum("i, j->ij", state_vector, state_vector), wires)
 
+    def _measure_expectation(self, tableau_simulator, meas_op, stim):
+        """Measure the expectation value with respect to the state of simulator device."""
+        # Get the observable for the expectation value measurement
+        meas_obs = qml.operation.convert_to_opmath(meas_op.obs)
+        coeffs, paulis = self._convert_op_to_linear_comb(meas_obs, coeffs=[], paulis=[])
+
+        expecs = qml.math.zeros_like(coeffs)
+        for idx, (pauli, wire) in enumerate(paulis):
+            pauli_term = ["I"] * max(np.max(list(wire)) + 1, tableau_simulator.num_qubits)
+            for op, wr in zip(pauli, wire):
+                pauli_term[wr] = op
+            stim_pauli = stim.PauliString("".join(pauli_term))
+            expecs[idx] = tableau_simulator.peek_observable_expectation(stim_pauli)
+
+        return qml.math.dot(coeffs, expecs)
+
+    def _measure_expectation_samples(self, stim_circuit, meas_op, num_shots, sample_seed):
+        """Measure the expectation value with respect to samples from simulator device."""
+        # Get the observable for the expectation value measurement
+        samples, coeffs = self._measure_observable_sample(
+            meas_op, stim_circuit, num_shots, sample_seed
+        )
+        expecs = [
+            qml.math.mean(qml.math.power([-1] * num_shots, qml.math.sum(sample, axis=1)))
+            for sample in samples
+        ]
+
+        return qml.math.dot(coeffs, expecs)
+
+    def _measure_variance(self, tableau_simulator, meas_op, stim):
+        """Measure the variance with respect to the state of simulator device."""
+        meas_obs = qml.operation.convert_to_opmath(meas_op.obs)
+        with qml.operation.use_new_opmath():
+            meas_obs1 = meas_obs.simplify()
+            meas_obs2 = (meas_obs1**2).simplify()
+
+        # use the naive formula for variance, i.e., Var(Q) = ⟨𝑄^2⟩−⟨𝑄⟩^2
+        return (
+            self._measure_expectation(tableau_simulator, ExpectationMP(meas_obs2), stim)
+            - self._measure_expectation(tableau_simulator, ExpectationMP(meas_obs1), stim) ** 2
+        )
+
+    # pylint: disable=protected-access
     @staticmethod
-    def _measure_purity(meas_op, circuit):
-        """Measure the purity of the state of simulator device"""
-        if circuit.wires != meas_op.wires:
-            raise NotImplementedError(
-                "default.clifford doesn't support measuring the purity of a subset of wires at the moment."
+    def _measure_stabilizer_vn_entropy(stabilizer, wires, log_base=None):
```

Comment:
Does stim have any similar method?

Reply:
No, stim doesn't have a similar method. However, there is a discussion regarding entropy [here](https://quantumcomputing.stackexchange.com/questions/16718/measuring-entanglement-entropy-using-a-stabilizer-circuit-simulator) essentially based on the A.1.d of this [paper](https://arxiv.org/pdf/1901.08092.pdf) I use.